src/server/auth/server/oidc.go

```diff
@@ -0,0 +1,216 @@
+	if !ok {
+		return "", fmt.Errorf("did not have a valid state")
+	}
+	oauthToken := si.Token
+
+	// Use the token passed in as our token source
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{
+			AccessToken: oauthToken,
+		},
+	)
+	userInfo, err := o.Provider.UserInfo(ctx, ts)
+	if err != nil {
+		return "", err
+	}
+	logrus.Infof("recovered user info with email: '%v'", userInfo.Email)
+	return userInfo.Email, nil
+}
+
+func (a *apiServer) handleExchange(w http.ResponseWriter, req *http.Request) {
+	ctx := context.Background()
+
+	var err error
+	cfg, sp := a.getOIDCSP()
+	if cfg == nil {
+		http.Error(w, "auth has no active config (either never set or disabled)", http.StatusConflict)
+		return
+	}
+	if sp == nil {
+		http.Error(w, "OIDC has not been configured or was disabled", http.StatusConflict)
+		return
+	}
+	if sp.Provider == nil {
+		logrus.Warn("cached provider was nil, but issuer info was present, so recovering")
+		sp.Provider, err = oidc.NewProvider(context.Background(), sp.Issuer)
+		if err != nil {
+			logrus.Errorf("could not find provider %v: %v", sp.Issuer, err)
+			http.Error(w, "OIDC Provider could not be found:", http.StatusConflict)
+			return
+		}
+	}
+
+	conf := &oauth2.Config{
+		ClientID:     sp.ClientID,
+		ClientSecret: sp.ClientSecret,
+		RedirectURL:  sp.RedirectURI,
+		Scopes:       []string{"openid", "email", "profile"},
+		Endpoint:     sp.Provider.Endpoint(),
+	}
+
+	code := req.URL.Query()["code"][0]
+	state := req.URL.Query()["state"][0]
+
+	logrus.Infof("session state: %v", state)
+
+	// Use the authorization code that is pushed to the redirect
+	tok, err := conf.Exchange(ctx, code)
+	if err != nil {
+		logrus.Errorf("failed to exchange code: %v", err)
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	logrus.Info("Exchanged OIDC code for token")
+
+	var verifier = sp.Provider.Verifier(&oidc.Config{ClientID: conf.ClientID})
+	// Extract the ID Token from OAuth2 token.
+	rawIDToken, ok := tok.Extra("id_token").(string)
+	if !ok {
+		logrus.Errorf("missing id token")
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	logrus.Infof("raw ID Token: %v", rawIDToken)
+
+	// Parse and verify ID Token payload.
+	idToken, err := verifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		logrus.Errorf("could not verify token: %v", err)
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	si := stateInfoMap[state]
+	if idToken.Nonce != si.Nonce {
+		logrus.Errorf("expected nonces to match, instead set nonce %v but got nonce %v", si.Nonce, idToken.Nonce)
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+	logrus.Infof("nonce is %v", idToken.Nonce)
+
+	si.Token = tok.AccessToken
+	logrus.Infof("saving state %v with access token %v", state, tok.AccessToken)
+	stateInfoMap[state] = si
+
+	// let the CLI know that we've successfully exchanged the code, and verified the token
+	tokenChan <- tokenInfo{token: tok.AccessToken, err: err}
```

Comment:
token chan should be part of state map

then don't need to clean-up, and protects from races